scipy/stats/_continuous_distns.py

```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            umax = np.sqrt(self._pdf(m, p, b))
+            s = sc.kv(p + 2, b) / sc.kv(p, b)
+            vmax = np.sqrt(self._pdf(self._mode(p + 2, b), p + 2, b) * s)
+            rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
+                                     umax=umax, vmin=0, vmax=vmax, c=0,
+                                     random_state=rndm)
+        else:
+            # new algorithm in [2]
+            x0 = b / (1 - p)
+            xs = np.max((x0, 2 / b))
+            k1 = self._pdf(m, p, b)
+            A1 = k1 * x0
+            if x0 < 2 / b:
+                k2 = np.exp(-b)
+                if p > 0:
+                    A2 = k2 * ((2 / b)**p - x0**p) / p
+                else:
+                    A2 = k2 * np.log(2 / b**2)
+            else:
+                k2, A2 = 0, 0
+            k3 = xs**(p - 1)
+            A3 = 2 * k3 * np.exp(-xs * b / 2) / b
+            A = A1 + A2 + A3
+
+            size1d = tuple(np.atleast_1d(sz))
+            N = np.prod(size1d)  # number of rvs needed, reshape upon return
+
+            # start sampling
+            rng = self._random_state
+            x = np.zeros(N)
+            simulated = 0
+            # [2]: rejection constant is < 2.73; so expected runtime is finite
+            while True:
+                k = N - simulated
+                h, rvs = np.zeros(k), np.zeros(k)
+                # simulate uniform rvs on [x1, x2] and [0, y2]
+                u1 = rng.random_sample(size=k)
+                v1 = A * rng.random_sample(size=k)
+                cond1 = v1 <= A1
+                cond2 = np.logical_not(cond1) & (v1 <= A1 + A2)
+                cond3 = np.logical_not(cond1 | cond2)
+                # subdomain (0, x0)
+                rvs[cond1] = x0 * v1[cond1] / A1
+                h[cond1] = k1
+                # subdomain (x0, 2 / b)
+                if p > 0:
+                    rvs[cond2] = (x0**p + (v1[cond2] - A1) * p / k2)**(1 / p)
+                else:
+                    rvs[cond2] = b * np.exp((v1[cond2] - A1) * np.exp(b))
+                h[cond2] = k2 * rvs[cond2]**(p - 1)
+                # subdomain (xs, infinity)
+                z = np.exp(-xs * b / 2) - b * (v1[cond3] - A1 - A2) / (2 * k3)
+                rvs[cond3] = -2 / b * np.log(z)
+                h[cond3] = k3 * np.exp(-rvs[cond3] * b / 2)
+                # apply rejection method
+                accept = (u1 * h <= self.pdf(rvs, p, b))
+                num_accept = sum(accept)
+                if num_accept > 0:
+                    take = min(num_accept, N - simulated)
+                    x[simulated:(simulated + take)] = rvs[accept][0:take]
+                    simulated += take
+                if simulated >= N:
+                    rvs = np.reshape(x, size1d)
+                    break
+
+        if invert_res:
+            rvs = 1 / rvs
+        return rvs
+
+    def _mode(self, p, b):
+        return b / (np.sqrt((p - 1)**2 + b**2) + 1 - p)
```

Comment:
This is the formula appropriate for p < 1, but it has problems when p > 1 and b is small.  The problem shows up in a call such as `geninvgauss.rvs(1.5, 1e-8)`, which generates the warnings shown below and fails to generate samples.

The problem is that when p > 1, and b is small, the formula for the mode suffers from catastrophic loss of precision.  It looks like this can be fixed by selecting the formula for the mode (equation (2) in the paper)
based on p's size relative to 1:

    def _mode(self, p, b):
        if p < 1:
            m = b / (np.sqrt((p - 1)**2 + b**2) + 1 - p)
        else:
            m = (np.sqrt((1 - p)**2 + b**2) - (1 - p)) / b
        return m

Here's the result of calling `geninvgauss.rvs(1.5, 1e-8)`:

```
In [34]: geninvgauss.rvs(1.5, 1e-8)                                                                          
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3546: RuntimeWarning: divide by zero encountered in double_scalars
  return b / (np.sqrt((p - 1)**2 + b**2) + 1 - p)
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3461: RuntimeWarning: invalid value encountered in double_scalars
  p1 = a1 - a2**2 / 3
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3462: RuntimeWarning: invalid value encountered in double_scalars
  q1 = 2 * a2**3 / 27 - a2 * a1 / 3 + m
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py:3384: RuntimeWarning: invalid value encountered in subtract
  lambda x, p, b, c: c + (p-1)*np.log(x) - b*(x+1/x)/2,
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-34-6c4504cd3606> in <module>
----> 1 geninvgauss.rvs(1.5, 1e-8)

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_distn_infrastructure.py in rvs(self, *args, **kwds)
    951         # by _rvs().
    952         self._size = size
--> 953         vals = self._rvs(*args)
    954 
    955         vals = vals * scale + loc

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py in _rvs(self, p, b)
   3395             out = self._rvs_scalar(p, b, self._size)
   3396         elif p.size == 1 and b.size == 1:
-> 3397             out = self._rvs_scalar(p.item(), b.item(), self._size)
   3398         else:
   3399             # When this method is called, self._size will be a (possibly empty)

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py in _rvs_scalar(self, p, b, numsamples)
   3471             rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
   3472                                      umax=umax, vmin=vmin, vmax=vmax, c=m,
-> 3473                                      random_state=rndm)
   3474         elif b >= min(0.5, 2 * np.sqrt(1 - p) / 3):
   3475             # ratio of uniforms without mode shift

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0_-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_rvs_sampling.py in rvs_ratio_uniforms(pdf, umax, vmin, vmax, size, c, random_state)
    166                    "to work for the provided parameters. Please check the "
    167                    "pdf and the bounds.".format(i*N))
--> 168             raise RuntimeError(msg)
    169         i += 1

RuntimeError: Not a single random variate could be generated in 50000 attempts. The ratio of uniforms method does not appear to work for the provided parameters. Please check the pdf and the bounds.
```

Reply:
you are right, i incorrectly assumed the version I had works for both cases, fixed now

Reply:
and it checked it: `geninvgauss.rvs(1.5, 1e-8)` returns a value immediately in this extreme case
`geninvgauss(1.5, 1e-8).stats() # (array(3.e+08), array(6.e+16))